src/rdkafka_transport.c

```diff
@@ -877,6 +1014,147 @@ int rd_kafka_transport_ssl_ctx_init (rd_kafka_t *rk,
+                            "Calling callback to retrieve client's private key password");
+                        len = rk->rk_conf.ssl.cert_retrieve_cb(RD_KAFKA_CERTIFICATE_PRIVATE_KEY_PASS, &buffer, rk->rk_conf.opaque);
+                        if (len) {
+                            rd_kafka_dbg(rk, SECURITY, "SSL",
+                                "Retrieved client's private key password %lu bytes",
+                                len);
+                            EVP_PKEY* pkey;
+                            X509 *cert;
+                            STACK_OF(X509) *ca = NULL;
+                            if (!PKCS12_parse(p12, (const char*)buffer, &pkey, &cert, &ca)) {
+                                rd_snprintf(errstr, errstr_size,
+                                    "Error reading PKCS#12");
+
+                                PKCS12_free(p12);
+                                BIO_free(bio);
+                                EVP_PKEY_free(pkey);
+                                X509_free(cert);
+                                if (ca != NULL)
+                                    sk_X509_pop_free(ca, X509_free);
+
+                                goto fail;
+                            }
+                            else {
+                                if (ca != NULL)
+                                    sk_X509_pop_free(ca, X509_free);
+
+                                X509_free(cert);
+
+                                rd_kafka_dbg(rk, SECURITY, "SSL",
+                                    "Setting client private key");
+
+                                r = SSL_CTX_use_PrivateKey(ctx, pkey);
+                                EVP_PKEY_free(pkey);
+
+                                if (r != 1) {
+                                    PKCS12_free(p12);
+                                    BIO_free(bio);
+
+                                    goto fail;
+                                }
+
+                                rd_kafka_dbg(rk, SECURITY, "SSL",
+                                    "Checking client private key");
+
+                                r = SSL_CTX_check_private_key(ctx);
+                                if (r != 1) {
+                                    PKCS12_free(p12);
+                                    BIO_free(bio);
+
+                                    goto fail;
+                                }
+                            }
+                        }
+                        
+                        PKCS12_free(p12);
+                    } else
+                        rd_snprintf(errstr, errstr_size,
+                            "d2i_PKCS12_bio failed");
+                    
+                    BIO_free(bio);
+                }
+                else
+                    rd_snprintf(errstr, errstr_size,
+                        "Failed to initialize BIO");
+        }
+        else
+            rd_kafka_dbg(rk, SECURITY, "SSL",
+                "Callback failed to return valid private key certificate");
```

Comment:
If this fails and sets errstr it also needs to `goto fail` to bug out, otherwise the errstr will be ignored.